### DISCUSSION TITLE
Validation API duplicates Markers

### DISCUSSION THREAD

**Discussion by rsoika** [eclipse-glsp/glsp#1281] [Q&A] [answered]
I am implementing the new Validation API in my project. We have [this discussion](https://github.com/eclipse-glsp/glsp/discussions/1280) regarding BATCH and LIVE validation. I think I understand the concepts how to validate now pretty well. 

But I observe another issue. It looks like GLSP caches - or not clears - marker lists.

In the following example I load a model with two errors. During load the 1st LIVE validation is triggered and the markers are displayed immediately on the elements and in the problem view.
When I connect the two elements (2nd LIVE validation) the markers are removed and the problem list becomes empty. This is wonderful!

![Peek 2024-03-03 09-36](https://github.com/eclipse-glsp/glsp/assets/1339124/404dfbfb-af4f-44b0-9f35-8fb1b24d8bdf)

In the next example, I load the same model (1st LIVE Validation) and click on the validation button in the Tool Palette (1st BATCH Validation). And now the markers are duplicated as you can see in the problem view. Even if I connect the two elements (2nd LIVE validation) only 2 of the 4 markers are cleared. So it looks like my diagram still have problems.
After I click once again on the Validation icon in the Tool Palette (2nd BATCH validation) finally all markers are cleared.

![Peek 2024-03-03 09-37](https://github.com/eclipse-glsp/glsp/assets/1339124/fa077fa5-b410-4e88-80df-93bd111840e3)

This looks to me that internally there are two marker lists hold by GLSP. For my understanding a BATCH Validation should clear all existing markers before.

**Comment by tortmayr** [comment]
Hi @rsoika,
sorry for the delayed response. 
I don't think that there is anything specifically wrong with your implementation.
It looks like per default when receiving a new set of validation markers the client only clears existing markers with the same validation reason. (https://github.com/eclipse-glsp/glsp-client/blob/645e72fe0e157729e056b93d9a3faa3df61fa1b2/packages/client/src/features/validation/validate.ts#L56)

I don't know the exact reasoning behind that maybe @planger has more insights on that topic.

**Reply by planger** [reply, marked as answer]
Yes, the idea behind that is to have a distinct set of rules for live validation and batch validation. This way, each validation rule is executed only once for the scope you want to run the validation (live or batch). Thus, when a batch validation is triggered, all markers of previous batch validations are cleared. When a live validation is triggered, all markers of the previous live validation are cleared. But they do not impact each other.

Otherwise, if an adopter doesn't make sure to invoke all live validation rules also from the batch validation, we would loose markers.
If the framework would enforce this (ie running all live validation rules on batch validation request), we would essentially do some unnecessary work by default, as the live validation results are per implementation always up to date anyway.

So I'd suggest to not run the live validation checks in the course of a batch validation, but only the expensive additional rules.

**Reply by rsoika** [reply]
@planger , OK understood, and in deed I have really a use for this too:

In our [Open-BPMN-Meta Project](https://github.com/imixs/open-bpmn/tree/master/open-bpmn.metamodel) we have implemented a general core-validation mechanism that is completely independent from GLSP. This validation mechanism is called in LIVE-Mode. We check only simple BPMN rules on a concrete element, like:

 - 'start event must not have an incoming flow' 
 - 'end event must not have an outgoing flow'
 - 'gateway must have in- and outgoing flows'
 - ....

This validation is very fast and uses an internal caching mode bound to the meta-model. 

And on the other side in Open-BPMN we have an extension mechanism hat allows adopters to extend the behavior for BPMN elements. We use this in our Imixs-BPMN Version. And for Imixs-Workflow models need to be validated in a more complex way. For example each event-element has an public ID that need to be unique in the context of the associated Task Element. This is not a general BPMN rule but specific for Imixs-Worklow. The validation is very complex - and so the BATCH Validation is the perfect place for that kind of rules!

And so it shows once again that GLSP is simply the perfect framework :-)
